dvc/fs/azure.py

```diff
@@ -68,158 +63,108 @@ def _az_config(self):
-        for blob in container_client.list_blobs(name_starts_with=prefix):
-            yield blob.name
+    def exists(self, path_info, use_dvcignore=False):
+        return self.client.exists(self._with_bucket(path_info))
 
-    def walk_files(self, path_info, **kwargs):
-        if not kwargs.pop("prefix", False):
-            path_info = path_info / ""
-        for fname in self._list_paths(
-            path_info.bucket, path_info.path, **kwargs
-        ):
-            if fname.endswith("/"):
-                continue
+    def _strip_bucket(self, entry):
+        _, entry = entry.split("/", 1)
+        return entry
 
-            yield path_info.replace(path=fname)
+    def _strip_buckets(self, entries, detail, prefix=None):
+        for entry in entries:
+            if detail:
+                entry["name"] = self._strip_bucket(entry["name"])
+            else:
+                entry = self._strip_bucket(
+                    f"{prefix}/{entry}" if prefix else entry
+                )
+            yield entry
 
     def ls(
         self, path_info, detail=False, recursive=False
     ):  # pylint: disable=arguments-differ
-        assert recursive
+        path = self._with_bucket(path_info)
+        if recursive:
+            for root, _, files in self.client.walk(path, detail=detail):
+                if detail:
+                    files = files.values()
+                yield from self._strip_buckets(files, detail, prefix=root)
+            return None
 
-        container_client = self.blob_service.get_container_client(
-            path_info.bucket
-        )
-        for blob in container_client.list_blobs(
-            name_starts_with=path_info.path
-        ):
-            if detail:
-                yield {
-                    "type": "file",
-                    "name": blob.name,
-                    "size": blob.size,
-                    "etag": blob.etag.strip('"'),
-                }
-            else:
-                yield blob.name
+        yield from self._strip_buckets(self.ls(path, detail=detail), detail)
 
-    def remove(self, path_info):
-        if path_info.scheme != self.scheme:
-            raise NotImplementedError
+    def walk_files(self, path_info, **kwargs):
+        for file in self.ls(path_info, recursive=True):
+            yield path_info.replace(path=file)
 
-        logger.debug(f"Removing {path_info}")
-        self.blob_service.get_blob_client(
-            path_info.bucket, path_info.path
-        ).delete_blob()
+    def remove(self, path_info):
+        self.client.delete(self._with_bucket(path_info))
 
     def info(self, path_info):
-        blob_client = self.blob_service.get_blob_client(
-            path_info.bucket, path_info.path
-        )
-        properties = blob_client.get_blob_properties()
-        return {
-            "type": "file",
-            "size": properties.size,
-            "etag": properties.etag.strip('"'),
-        }
+        self.client.invalidate_cache()
```

Comment:
The `invalidate_cache` calls here and in the download() are still on my radar, but I believe this might be related to a bug in the upstream (still didn't got the standalone reproducer, but the tests fail without them since the dircache somehow holds the containerless-path but the `_ls_from_cache` function tries to match it with the path that also includes the container).